File: backend-rust/src/runtime/validation.rs
use super::json::value_profile;
use serde_json::{Value, json};
use std::collections::{BTreeSet, HashSet};

pub(super) fn profile(root: &Value) -> Result<(), &'static str> {
    if !value_profile(root) {
        return Err("schema limits");
    }
    fn walk(
        root: &Value,
        node: &Value,
        seen: &HashSet<String>,
        depth: usize,
    ) -> Result<(), &'static str> {
        if depth > 64 {
            return Err("schema recursion");
        };
        if node.is_boolean() {
            return Ok(());
        };
        let m = node.as_object().ok_or("invalid schema")?;
        let supported = "$schema $id $defs $ref type properties required additionalProperties enum const minimum maximum exclusiveMinimum exclusiveMaximum minLength maxLength minItems maxItems items minProperties maxProperties allOf anyOf oneOf title description default examples deprecated readOnly writeOnly format pattern";
        if root["$schema"] == "http://json-schema.org/draft-07/schema#"
            && m.contains_key("$ref")
            && m.keys().any(|k| {
                !"$ref title description default examples deprecated readOnly writeOnly format"
                    .split_whitespace()
                    .any(|allowed| allowed == k)
            })
        {
            return Err("draft-07 ref siblings outside common subset");
        }
        for (k, v) in m {
            if !supported.split_whitespace().any(|s| s == k) {
                return Err("unsupported schema");
            };
            match k.as_str() {
                "pattern" => {
                    if !WOKU_PATTERNS.contains(&v.as_str().ok_or("invalid pattern")?) {
                        return Err("unsupported pattern");
                    }
                }
                "$schema" => {
                    if v != "https://json-schema.org/draft/2020-12/schema"
                        && v != "http://json-schema.org/draft-07/schema#"
                    {
                        return Err("schema dialect");
                    }
                }
                "$ref" => {
                    let reference = v.as_str().ok_or("schema reference")?;
                    if !reference.starts_with("#/") || seen.contains(reference) {
                        return Err("schema reference");
                    };
                    let target = root.pointer(&reference[1..]).ok_or("schema reference")?;
                    let mut next = seen.clone();
                    next.insert(reference.into());
                    walk(root, target, &next, depth + 1)?;
                }
                "properties" | "$defs" => {
                    for child in v.as_object().ok_or("schema map")?.values() {
                        walk(root, child, seen, depth + 1)?
                    }
                }
                "items" | "additionalProperties" => walk(root, v, seen, depth + 1)?,
                "allOf" | "anyOf" | "oneOf" => {
                    for child in v.as_array().ok_or("schema branches")? {
                        walk(root, child, seen, depth + 1)?
                    }
                }
                _ => {}
            }
        }
        Ok(())
    }
    walk(root, root, &HashSet::new(), 0)
}
pub(super) fn compile(schema: &Value) -> Result<jsonschema::Validator, &'static str> {
    profile(schema)?;
    let mut compiled = schema.clone();
    normalize_patterns(&mut compiled);
    jsonschema::options()
        .with_draft(jsonschema::Draft::Draft202012)
        .should_validate_formats(false)
        .build(&compiled)
        .map_err(|_| "invalid schema")
}
fn escaped(s: &str) -> String {
    s.replace('~', "~0").replace('/', "~1")
}
pub(super) fn validate(schema: &jsonschema::Validator, args: &Value) -> (Vec<Value>, Vec<String>) {
    if !value_profile(args) || !args.is_object() {
        return (
            vec![
                json!({"code":"unsupported_value","instance_path":"","schema_path":"#","message_key":"toolgate.validation.unsupported_value"}),
            ],
            vec![],
        );
    }
    let mut issues = Vec::new();
    let mut missing = BTreeSet::new();
    for error in schema.iter_errors(args) {
        let path = error.instance_path().to_string();
        let schema_path = error.schema_path().to_string();
        let keyword = schema_path.rsplit('/').next().unwrap_or("");
        let code = match keyword {
            "required" => "required",
            "type" => "type",
            "enum" => "enum",
            "const" => "const",
            "additionalProperties" => "additional_property",
            "minimum" | "exclusiveMinimum" => "minimum",
            "maximum" | "exclusiveMaximum" => "maximum",
            "minLength" | "maxLength" => "length",
            "items" | "minItems" | "maxItems" => "items",
            "oneOf" => "one_of",
            "anyOf" => "any_of",
            "allOf" => "all_of",
            _ => "unsupported_value",
        };
        let instance =
            if let jsonschema::error::ValidationErrorKind::Required { property } = error.kind() {
                let property = property.as_str().unwrap_or("");
                let p = format!("{path}/{}", escaped(property));
                missing.insert(p.clone());
                p
            } else {
                path
            };
        issues.push(json!({"code":code,"instance_path":instance,"schema_path":format!("#{schema_path}"),"message_key":format!("toolgate.validation.{code}")}));
    }
    issues.sort_by_key(|v| {
        (
            v["instance_path"].as_str().unwrap_or("").to_owned(),
            v["schema_path"].as_str().unwrap_or("").to_owned(),
            v["code"].as_str().unwrap_or("").to_owned(),
        )
    });
    issues.dedup();
    issues.truncate(128);
    (issues, missing.into_iter().take(128).collect())
}

// Finite catalog vocabulary; arbitrary backtracking expressions remain rejected.
const WOKU_PATTERNS: &[&str] = &[
    r"^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$",
    r"\S",
    r"^(?=.*[0-9])[+]?[0-9()\-\s.]+$",
    r"^[a-f0-9]{24}$",
    r"^[0-9a-fA-F]{24}$",
    r"\D",
];
fn normalize_patterns(value: &mut Value) {
    if let Some(map) = value.as_object_mut() {
        if let Some(pattern) = map.get_mut("pattern")
            && let Some(text) = pattern.as_str()
        {
            let spaces =
                r"\x09-\x0d\x20\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000\ufeff";
            let translated = text
                .replace(r"\D", "[^0-9]")
                .replace(r"\S", &format!("[^{spaces}]"))
                .replace(r"\s", &format!("[{spaces}]"))
                .replace(".*", r"[^\n\r\u2028\u2029]*");
            *pattern = Value::String(translated);
        }
        // Traverse schema positions only; annotation/example objects are data.
        for (key, child) in map.iter_mut() {
            match key.as_str() {
                "properties" | "$defs" => {
                    if let Some(entries) = child.as_object_mut() {
                        for node in entries.values_mut() {
                            normalize_patterns(node);
                        }
                    }
                }
                "items" | "additionalProperties" => normalize_patterns(child),
                "allOf" | "anyOf" | "oneOf" => {
                    if let Some(entries) = child.as_array_mut() {
                        for node in entries {
                            normalize_patterns(node);
                        }
                    }
                }
                _ => {}
            }
        }
    }
}
#[cfg(test)]
mod woku_pattern_tests {
    use super::*;
    #[test]
    fn draft7_ref_validation_siblings_rejected() {
        let schema = json!({"$schema":"http://json-schema.org/draft-07/schema#","type":"object","$defs":{"s":{"type":"string"}},"properties":{"s":{"$ref":"#/$defs/s","minLength":3}}});
        assert!(compile(&schema).is_err());
    }
    #[test]
    fn ecmascript_parity() {
        let cases: Value =
            serde_json::from_str(include_str!("../../../shared/fixtures/woku-patterns.json"))
                .expect("fixture");
        for case in cases.as_array().expect("cases") {
            let schema = json!({"type":"object","properties":{"value":{"type":"string","pattern":case["pattern"]}}});
            let validator = compile(&schema).expect("pattern compiled");
            assert_eq!(
                validator.is_valid(&json!({"value":case["value"]})),
                case["valid"].as_bool().expect("bool"),
                "{case}"
            );
        }
    }
}

File: backend-go/internal/core/validation.go
package core

import (
	_ "embed"
	"errors"
	"fmt"
	"sort"
	"strings"

	js "github.com/santhosh-tekuri/jsonschema/v6"
	"github.com/santhosh-tekuri/jsonschema/v6/kind"
)

//go:embed api.generated.json
var apiBytes []byte

type denyLoader struct{}

func (denyLoader) Load(string) (any, error) {
	return nil, errors.New("external schema loading disabled")
}
func compiler() *js.Compiler {
	c := js.NewCompiler()
	c.DefaultDraft(js.Draft2020)
	c.UseLoader(denyLoader{})
	return c
}
func wireSchemas() (map[string]*js.Schema, error) {
	raw, err := ParseJSON(apiBytes)
	if err != nil {
		return nil, err
	}
	object, ok := raw.(map[string]any)
	if !ok {
		return nil, errJSON
	}
	defs, ok := object["$defs"].(map[string]any)
	if !ok {
		return nil, errJSON
	}
	c := compiler()
	c.AssertFormat()
	if err = c.AddResource("https://toolgate.invalid/api", raw); err != nil {
		return nil, err
	}
	result := map[string]*js.Schema{}
	for name := range defs {
		schema, e := c.Compile("https://toolgate.invalid/api#/$defs/" + name)
		if e != nil {
			return nil, e
		}
		result[name] = schema
	}
	return result, nil
}

type Issue struct {
	Code         string `json:"code"`
	InstancePath string `json:"instance_path"`
	SchemaPath   string `json:"schema_path"`
	MessageKey   string `json:"message_key"`
}

func pointer(parts []string) string {
	result := ""
	for _, p := range parts {
		result += "/" + strings.ReplaceAll(strings.ReplaceAll(p, "~", "~0"), "/", "~1")
	}
	return result
}
func compileTool(schema map[string]any) (*js.Schema, error) {
	if err := schemaProfile(schema); err != nil {
		return nil, err
	}
	c := compiler()
	c.UseRegexpEngine(compileWokuPattern)
	if err := c.AddResource("https://toolgate.invalid/tool", schema); err != nil {
		return nil, err
	}
	return c.Compile("https://toolgate.invalid/tool")
}

// ValidateArguments is also used by deterministic unit fixtures.
func ValidateArguments(schema map[string]any, args map[string]any) ([]Issue, []string, error) {
	compiled, e := compileTool(schema)
	if e != nil {
		return nil, nil, e
	}
	issues, missing := validateCompiled(compiled, args)
	return issues, missing, nil
}
func validateCompiled(compiled *js.Schema, args map[string]any) ([]Issue, []string) {
	issues := []Issue{}
	missing := []string{}
	raw, e := Canonical(args)
	if e != nil || len(raw) > 65536 || !valueProfile(args, 0) {
		return []Issue{{"unsupported_value", "", "#", "toolgate.validation.unsupported_value"}}, missing
	}
	err := compiled.Validate(args)
	if err == nil {
		return issues, missing
	}
	var validation *js.ValidationError
	if !errors.As(err, &validation) {
		return []Issue{{"unsupported_value", "", "#", "toolgate.validation.unsupported_value"}}, missing
	}
	var walk func(*js.ValidationError)
	walk = func(v *js.ValidationError) {
		keyword := v.ErrorKind.KeywordPath()
		key := ""
		if len(keyword) > 0 {
			key = keyword[0]
		}
		location := "#"
		if at := strings.Index(v.SchemaURL, "#"); at >= 0 {
			location = v.SchemaURL[at:]
		}
		schemaPath := location + pointer(keyword)
		code := key
		switch key {
		case "additionalProperties":
			code = "additional_property"
		case "minLength", "maxLength":
			code = "length"
		case "minItems", "maxItems", "items":
			code = "items"
		case "exclusiveMinimum":
			code = "minimum"
		case "exclusiveMaximum":
			code = "maximum"
		case "pattern", "minProperties", "maxProperties":
			code = "unsupported_value"
		case "oneOf":
			code = "one_of"
		case "anyOf":
			code = "any_of"
		case "allOf":
			code = "all_of"
		}
		if key == "anyOf" || key == "oneOf" {
			issues = append(issues, Issue{code, pointer(v.InstanceLocation), schemaPath, "toolgate.validation." + code})
			return
		}
		if len(v.Causes) > 0 {
			for _, cause := range v.Causes {
				walk(cause)
			}
			return
		}
		if required, ok := v.ErrorKind.(*kind.Required); ok {
			for _, name := range required.Missing {
				path := pointer(append(append([]string{}, v.InstanceLocation...), name))
				issues = append(issues, Issue{"required", path, schemaPath, "toolgate.validation.required"})
				missing = append(missing, path)
			}
			return
		}
		if code == "" {
			code = "unsupported_value"
		}
		issues = append(issues, Issue{code, pointer(v.InstanceLocation), schemaPath, "toolgate.validation." + code})
	}
	walk(validation)
	sort.Slice(issues, func(i, j int) bool {
		a, b := issues[i], issues[j]
		return a.InstancePath+"\x00"+a.SchemaPath+"\x00"+a.Code < b.InstancePath+"\x00"+b.SchemaPath+"\x00"+b.Code
	})
	unique := issues[:0]
	for _, i := range issues {
		if len(unique) == 0 || unique[len(unique)-1] != i {
			unique = append(unique, i)
		}
	}
	issues = unique
	if len(issues) > 128 {
		issues = issues[:128]
	}
	sort.Strings(missing)
	if len(missing) > 128 {
		missing = missing[:128]
	}
	return issues, missing
}
func schemaProfile(root map[string]any) error {
	if !valueProfile(root, 0) {
		return errors.New("schema limits")
	}
	raw, e := Canonical(root)
	if e != nil || len(raw) > 65536 {
		return errors.New("schema limits")
	}
	supported := map[string]bool{}
	for _, k := range strings.Fields("$schema $id $defs $ref type properties required additionalProperties enum const minimum maximum exclusiveMinimum exclusiveMaximum minLength maxLength minItems maxItems items minProperties maxProperties allOf anyOf oneOf title description default examples deprecated readOnly writeOnly format pattern") {
		supported[k] = true
	}
	var walk func(any, map[string]bool, int) error
	walk = func(node any, seen map[string]bool, depth int) error {
		if depth > 64 {
			return errors.New("schema recursion")
		}
		if _, ok := node.(bool); ok {
			return nil
		}
		obj, ok := node.(map[string]any)
		if !ok {
			return errors.New("invalid schema")
		}
		if root["$schema"] == "http://json-schema.org/draft-07/schema#" {
			if _, hasRef := obj["$ref"]; hasRef {
				for key := range obj {
					if !contains(strings.Fields("$ref title description default examples deprecated readOnly writeOnly format"), key) {
						return errors.New("draft-07 ref siblings outside common subset")
					}
				}
			}
		}
		for k, value := range obj {
			if !supported[k] {
				return fmt.Errorf("unsupported schema keyword")
			}
			switch k {
			case "pattern":
				p, ok := value.(string)
				if !ok || !wokuPatterns[p] {
					return errors.New("unsupported pattern")
				}
			case "$schema":
				if value != "https://json-schema.org/draft/2020-12/schema" && value != "http://json-schema.org/draft-07/schema#" {
					return errors.New("schema dialect")
				}
			case "$ref":
				ref, ok := value.(string)
				if !ok || !strings.HasPrefix(ref, "#/") || seen[ref] {
					return errors.New("schema reference")
				}
				target := any(root)
				for _, part := range strings.Split(ref[2:], "/") {
					m, ok := target.(map[string]any)
					if !ok {
						return errors.New("schema reference")
					}
					target = m[strings.ReplaceAll(strings.ReplaceAll(part, "~1", "/"), "~0", "~")]
				}
				next := map[string]bool{}
				for key, v := range seen {
					next[key] = v
				}
				next[ref] = true
				if err := walk(target, next, depth+1); err != nil {
					return err
				}
			case "properties", "$defs":
				m, ok := value.(map[string]any)
				if !ok {
					return errors.New("schema map")
				}
				for _, child := range m {
					if err := walk(child, seen, depth+1); err != nil {
						return err
					}
				}
			case "items", "additionalProperties":
				if err := walk(value, seen, depth+1); err != nil {
					return err
				}
			case "allOf", "anyOf", "oneOf":
				list, ok := value.([]any)
				if !ok {
					return errors.New("schema branches")
				}
				for _, child := range list {
					if err := walk(child, seen, depth+1); err != nil {
						return err
					}
				}
			}
		}
		return nil
	}
	return walk(root, map[string]bool{}, 0)
}
